### ISSUE TITLE
Cafes UI

### ISSUE THREAD

**Issue by undercase** [textileio/photos#1163]
## Description
Add user interface for cafes and peers to the application.

**Comment by undercase** [comment]
I want to consolidate all my thinking on this, so here it goes. I've spent this past week working on and thinking about this issue. I've been able to go through all of the existing cafes code I could find and I think I have a good grasp of the cafes workflow. In the process of this work, I've had some success, ran into some problems, and I have a few questions.

### Questions

- Where should I make the API call to determine which cafes are nearest to the user? Right now I am doing the logic asynchronously using `textile-lb-api` in `componentDidMount()` just so it's there, but I think I should probably put it in a saga. If I do put it in a saga, I'm not sure what part of the state I should store the request/results/errors in.

- Where should I store the hardcoded cafe information? As far as I can tell, it's not possible to store it as JSON in the `.env`, and I'm not sure how `App/Config` works. Right now I just have it stored as a variable in the `CafesList` component file, where it is used.

- At what point in onboarding is the user registered with a cafe? In `App/features/account/sagas.ts`, `registerCafesIfNeeded()` is called whenever the node starts up, which executes `registerCafes()` if the user isn't registered with a cafe. This code uses the aforementioned `textile-lb-api` to find the nearest cafes and registers the user with one of those. This all makes sense to me, but that code relies on environment variables which do not exist in my `.env`. Despite this, I am still somehow registered with a cafe after onboarding? I'd really like to know how this is happening. My `.env` also features a number of environment variables which seem relevant, but are never used anywhere in the application (tested using `grep`):

```
RN_TEXTILE_CAFE_OVERRIDE="https://us-west-beta.textile.cafe"
...
RN_TEXTILE_CAFE_GATEWAY_URL="http://13.56.163.77:5050"
```

- The existing code for cafe registration uses cafe url for registration, and it seems like the latest SDK relies on them as well, but @sanderpick has mentioned to me that we want to use peer IDs for registration. It seems like the javascript api and command line interface have already switched to peer IDs. Should I go ahead and use urls? Or table this whole issue for now and wait to write new code until the sdk has been updated?

### Problems

- I originally planned to use `react-native-swipe-list-view` for the `Deregister` button on the cafes settings view, but that library has a lot of issues and seems like it's very poorly maintained (even though it has a ton of stars?). I'm just going to table that feature and put the deregister button in the cafe screen itself.

- I feel like this feature has honestly taken me longer than it should, and I really want to get at least a draft totally done in the next day or two. This isn't really a problem, just an observation.

### Success

I'm opening a draft PR to make it easy to see the code for these features:

- The cafes onboarding screen looks and works great, and in the process of designing it I expanded on the work from contact multi-select and made a nice `Checkbox` element.

- The cafes setting screen / individual cafe view is also functional, although it doesn't have the `Deregister` button like the mockup, and the `Services` list is hardcoded in as a placeholder (since the logic for assigning certain cafes to certain services isn't there).

tl;dr: The UI is done. I totally get the logic and API for the cafes "backend" but I'm not sure how to proceed with it. I'd like to leave open the possibility that I have totally misunderstood the existing code.

**Comment by asutula** [comment]
Nice, sounds like some great progress. I'll try to step through each of your bullets here:

- Sounds like we should create a redux "feature" (a subdir in our "features" dir) for cafe-related things. All the logic, state, and I/O for cafes should live in there. You'll be able to have some redux state that holds a "loading" state while a saga is calling the `textile-lb-api`, trigger that saga using an action when the "choose a cafe" screen is loading during onboarding. 
- I'm not sure what JSON data you're needed to store in the `.env`... The only constants required for registering a cafe are `RN_TEXTILE_LB_URL` or `RN_TEXTILE_CAFE_URL` (soon to be `RN_TEXTILE_CAFE_PEER_ID`). Any other information about previously registered cafes is persisted in the `go-textile` layer and is queried and stored in redux as needed.
- I'm not sure about you somehow registering a cafe even when the env vars don't exist, but... We will want to remove the automatic call to `registerCafesIfNeeded`. A call to `registerCafe` will only happen when the user explicitly chooses to register a cafe during onboarding or later on from the settings screens. The cafe to be registered will come either from our LB api or from the user manually entering the information. You can see the current valid env values in https://github.com/textileio/photos/blob/master/.env.example
- The version of the SDK that is used in your branch uses cafe URL, so you'll have to use URL. But an update to the SDK that is currently WIP switches to peer id. Once that's out, you can update your branch and start using peer id. It's a pretty simple switch, just type `.peer` instead of `.url`.
- I don't think your problems are very big problems. I'm all for avoiding 3rd party dependencies that are poor quality, even if it means a more simple or less ideal implementation for us.
- Work always takes longer than you think. This is a complicated one, don't worry about it!

Let's chat if you want, could be good to just make sure we're on the same page.

**Comment by sanderpick** [comment]
+1 to @asutula's points. There a lot going on here so thanks for dumping your thoughts! One more thing comes to mind:

- Do we want to use the "nearest cafe" technique at all anymore? Is the idea there to just tell them which one is closest, sort of like a recommendation?

**Comment by undercase** [comment]
Yeah, right now the UI is using the nearest cafe feature for a recommendation. It just says *Recommended* below the peer ID for the two cafes, like in the mockup.

**Comment by undercase** [comment]
Check out the draft PR #1175 to see the UI code. Working on building out the backend saga code following @asutula's suggestions!

**Comment by undercase** [comment]
This is all great, and I'm super clear about how to proceed. Making really strong progress today. Only thing-regarding the .env, sander and I agreed that we should store a hardcoded list of the textile cafes (e.g. us-west, us-east, etc.) somewhere and show that in onboarding, along with the option to enter your own peerId. We talked about storing this in the config (I assumed that meant the .env) but I couldn't find a place to put it, so it's just hardcoded into `CafesList.tsx` right now.

**Comment by sanderpick** [comment]
Makes sense to me!
